\chapter*{Introduction}
\addcontentsline{toc}{chapter}{Introduction}
This document is written based on a \href{http://phd.epfl.ch/thesistemplates}{Generic EPFL Thesis Template in LaTeX v. 5.57} and may be extended in the vignette section as it is distributed through GitHub to annotate the statanacoseq project and this chapter should contain the text elements that are used to build up a proper R vignette of the package. Meaning to give examples for the usage of the different indices and functions for biological interpretation with the code for repeating the \textit{in silico} experiments by applying the code lines step by step or by copy pasting the entire code for a sample experiment which raw data can be found in a public database. At the moment the vignette section just list the included functions and data, as well as a short description as in the package overview documentation.

\section{Aim}
We are aiming to set up a tool to automatically calculate and evaluate statistical properties of codon bias, especially implementing the calculation for different codon bias indices and for genes of organisms where its genome is sequenced for the most part. That means that we have at least a partially sequenced genome and the information on expressed sequences or proteins.
The most simple calculation of codon bias is the GC content of the third base of synonymous codons of a single gene or coding DNA sequence (CDS). This codon bias reflects neutral mechanisms like mutational selection for a high or low GC content in an organism. The CG content of the third coding base generally is around 40 percent in low GC-regions but may increase up to 80 percent in high GC-regions \cite{Galtier2001}. This \hyperlink{function:G3C}{GC3 value} can be easily calculated by the coding sequence alone and has been associated with overall GC content of a genome \cite{Muto1987}.

Such mechanisms based on a single gene can be revised by incorporating species independent information of general translation efficiency and RNA folding. Moreover, instead of analyzing a single nucleotide position alone (mononucleotides), information on cytosine (C)-methylation can be taken into account and calculation on dinucleotides (GC) for mammals or quadnucleotide (GATC) or pentnucleotides (CCWGG) which harbors a degenerate nucleotide position that can be A or T. The succession of codons (bicodons or even tricodons) may of course also be of interest but may only refine codon bias indexes where information on the whole genome or even for species is taken into account.

When we aim for introducing a little bit more sophisticated indexes, they will depend on additional information such as, the complete transcriptome, the complete genome, gene expression, tRNA availability and protein properties.

One aspect for some codon bias measurements is the definition of the optimal codon, thus the one with the most desirable quality. This information is used to measure the \hyperlink{function:Fop}{frequency of optimal codons (Fop)} or the P1 index, a measure of the influence of tRNA availability. Which of the codons is optimal can be manifoldly defined, such as predicted using the tRNA gene copy number as a indicator as tRNA levels correlate with gene copy number in general \cite{Chaney2015}. %(Chaney and Clark 2015)

\section{Teff genome}
\textit{Eragostis tef} is not only a grass but an important crop, that was neglected in scientific studies before the setup of the 'Teff Improvement Project" ten years ago at the University of Bern. Applied science generated improved plants that made it to the farmers in Ethiopia and research is ongoing to discover new break-through plants that show increased seed size, soil acidity, drought and herbic	ide tolerance along with the improved yield and to maintain lodging resistance without being semi-dwarf. A valuable resource for plants genetics is also the sequencing and annotation of the tetraploid genome of Teff and the evaluation of its properties \cite{cannarozzi2014genome}. Therefore, we used as sample organism here, sequences from this genome to examine what obstacles one encounters by working with a almost totally sequenced but not yet fully annotated genome. By learning from its codon usage, variations generated by technologies already used in the 1960 for screening of genetic variations of agricultural plants combined with DNA sequencing technologies may improve the selection criteria for optimal mutants used for back crossing. 

\begin{figure}[tb] 
\centering 
\includegraphics[width=0.9\columnwidth]{teff} 
\caption[\textit{Eragostis tef} growing in culture room]{Photography of model grass plants of species \textit{E. tef}}
\label{fig:teff} 
\end{figure}


\section{Biological aspects of Codon usage}
The ribosome is coordinating the translation from messenger RNA (mRNA) to proteins in the cells together with tranlsational factors and aminoacyl-transfer RNAs (tRNA) and guanosinetriphosphate as substrate. The efficiency of translation is thereby defined by the mRNA quantity and stability, but also by the composition (concinnity, order and the distribution of codons, the bioavailability of the mRNA and the translational factors and much more. The bias for using particular codons preferentially must be differentiated between a \textit{codon bias}, which tends to use those codons that will lead to an optimal use of tRNA for having an optimal cost-benefit factor between a fast and a accurate protein synthesis (including aminoacid misincorporations and protein malfolding) and a \textit{codon adaptation}, that is aiming to optimize the use of codons towards a enhanced translational efficiency and this is subject to a exchange trend towards more optimal codons.
For estimation of the codon bias we can compute codon usage bu measuring the bias to use adaptive codons for example with the \hyperlink{function:CAI}{Codon Adaptation Index (CAI)} that is implemented in our code. The \hyperlink{function:RA}{relative adaptiveness (RA)} has thereby be calculated form a separate set of sequences, usually from the usage of highly expressed genes (codon fitness values). For the moment that is the users task, but may be decided on gene expression data in the future in our package. In contrast, the \hyperlink{function:NEC}{Number of effective Codons (NEC)} in our package measures only the degree of deviation from an uniform codon usage based on basic statistics, regardless of which codons are overrepresented \cite{Suzuki2016}.
Moreover the package could be expanded in the future to compute additional indices such as the S value, that measures the strength of selection for codon adaptation (S), to perform within-group correspondence analysis (WCA) of codon usage, give information on replication strand skew analysis, the GC skew index (GCSI), single statistics (S) or PCA to estimate the relative selection strength within a genome.

\subsection{Genetic Code}
Here is shown a table with the standard genetic code (from verified ORF from \textit{E. teff}), generated with the first of the following functions:  
  \lstinputlisting[language=R, breaklines=true]{codes/geneticcode.R}
  \lstinputlisting[language=R, breaklines=true]{codes/CodonTable.R}

As the verified CDS from E. teff have their stop codon already chopped off, we don't count them in the codon table below.\\
\begin{table}
\begin{center}
{\tiny
\begin{tabular}{*{21}{l}}
\hline
\\
TTT & Phe & 231800	& 1.44\%	&	&
TCT & Ser & 231219	& 1.43\%	&	&
TAT & Tyr & 169761	& 1.05\%	&	&
TGT & Cys & 99497	& 0.62\%	&	&
\\
TTC & Phe & 363171	& 2.25\%	&	&
TCC & Ser & 236162	& 1.46\%	&	&
TAC & Tyr & 250318	& 1.55\%	&	&
TGC & Cys & 181252	& 1.12\%	&	&
\\
TTA & Leu & 101045	& 0.63\%	&	&
TCA & Ser & 234348	& 1.45\%	&	&
TAA & Stp & 0	& 0\%	&	&
TGA & Stp & 0	& 0\%	&	&
\\
TTG & Leu & 239454	& 1.48\%	&	&
TCG & Ser & 155902	& 0.97\%	&	&
TAG & Stp & 0	& 0\%	&	&
TGG & Trp & 197407	& 1.22\%	&	&
\\
\\
CTT & Leu & 289697	& 1.79\%	&	&
CCT & Pro & 221214	& 1.37\%	&	&
CAT & His & 182484	& 1.13\%	&	&
CGT & Arg & 106245	& 0.66\%	&	&
\\
CTC & Leu & 356286	& 2.21\%	&	&
CCC & Pro & 166959	& 1.03\%	&	&
CAC & His & 189362	& 1.17\%	&	&
CGC & Arg & 190954	& 1.18\%	&	&
\\
CTA & Leu & 119917	& 0.74\%	&	&
CCA & Pro & 230207	& 1.43\%	&	&
CAA & Gln & 208855	& 1.29\%	&	&
CGA & Arg & 82144	& 0.51\%	&	&
\\
CTG & Leu & 356616	& 2.21\%	&	&
CCG & Pro & 205050	& 1.27\%	&	&
CAG & Gln & 332504	& 2.06\%	&	&
CGG & Arg & 156854	& 0.97\%	&	&
\\
\\
ATT & Ile & 257976	& 1.60\%	&	&
ACT & Thr & 188031	& 1.16\%	&	&
AAT & Asn & 266263	& 1.65\%	&	&
AGT & Ser & 158479	& 0.98\%	&	&
\\
ATC & Ile & 308634	& 1.91\%	&	&
ACC & Thr & 210048	& 1.30\%	&	&
AAC & Asn & 304586	& 1.89\%	&	&
AGC & Ser & 246403	& 1.53\%	&	&
\\
ATA & Ile & 155768	& 0.96\%	&	&
ACA & Thr & 208202	& 1.29\%	&	&
AAA & Lys & 284060	& 1.76\%	&	&
AGA & Arg & 169467	& 1.05\%	&	&
\\
ATG & Met & 378621	& 2.35\%	&	&
ACG & Thr & 151392	& 0.94\%	&	&
AAG & Lys & 549104	& 3.40\%	&	&
AGG & Arg & 226367	& 1.40\%	&	&
\\
\\
GTT & Val & 289284	& 1.79\%	&	&
GCT & Ala & 345407	& 2.14\%	&	&
GAT & Asp & 431250	& 2.67\%	&	&
GGT & Gly & 246669	& 1.53\%&	&
\\
GTC & Val & 291192	& 1.80\%	&	&
GCC & Ala & 377503	& 2.34\%	&	&
GAC & Asp & 412788	& 2.56\%	&	&
GGC & Gly & 374322	& 2.32\%	&	&
\\
GTA & Val & 118050	& 0.73\%	&	&
GCA & Ala & 306736	& 1.90\%	&	&
GAA & Glu & 369243	& 2.29\%&	&
GGA & Gly & 250072	& 1.55\%	&	&
\\
GTG & Val & 361307	& 2.24\%	&	&
GCG & Ala & 289771	& 1.79\%	&	&
GAG & Glu & 580359	& 3.60\%	&	&
GGG & Gly & 227638	& 1.41\%	&	&
\\
\\
\hline
\end{tabular}
\caption[Genetic code table]{Table with standard genetic code}
\label{tab:sgencod}
}
\end{center}
\end{table}


