\chapter{Vignette}
This document is a vignette for using the R package statanacoseq:


\section{statanacoseq}


\section{Demos}
Fop \quad Computes the Frequency of Optimal Codons in the Eragostis tef Sample Set \\
NEC \quad Computes the Overall Number of Effective Codons in the Eragostis tef Sample Set \\
plotOCU \quad Plots the Codon Optimality Score for Each Aminoacid in the Eragostis tef Sample Set with Moving Averages \\


\section{Tests}
checkCDNA \quad \\
Fop \quad \\
NEC \quad \\
NucleotideContent \quad \\
species \quad \\

\section{Functions}

\subsection{Indices}
\hypertarget{function:CAI}{ComputeCAI}	\quad Codon Adaptation Index \\
ComputeCAIVector \quad 	Codon Adaptation Index Vector \\
ComputeCarboneRA	\quad Compute Carbone Relative Adaptiveness \\
\hypertarget{function:Fop}{ComputeFop}	\quad Frequency of Optimal Codons \\
ComputeGC3syn	\quad GC content 3rd position of synonymous codons	 \\	% need fixing
\hypertarget{function:NEC}{ComputeNEC}	\quad Effective number of codons \\							% I guess all functions need require(seqinr)
\hypertarget{function:RA}{RelativeAdaptiveness}	\quad Compute Relative Adaptiveness	 \\			% Handle warnings

\subsubsection{Intermediator functions for indices}
RSCU	Relative \quad Synonymous Codon Usage \\								% 

\subsubsection{Void indices}
ComputeCBI	\quad Codon Bias Index  \\
ScaledChiSquare	\quad Scaled Chi-Square  \\
SilentSiteComposition	\quad Base Composition at Silent Sites  \\

\subsubsection{Database Function from Darwin left untranscribed}
Entries	\quad Entries \\

\subsection{Helperfunctions}
anticodoncount	\quad Calculate Global Anticodon Usage \\
areopts	\quad Optimality Decision for Codons of Custom Genome \\
checkCDS	Checks \quad Integrity of CDS Sequence \\
CodonProb	\quad Codon Probability table \\
CodonProbabilities	\quad Codon Probabilities \\
CodonUsage	\quad Table for Codon Usage and Optimality of Codons \\
FindHighlyExpressedGenes	\quad Find Highly Expressed Genes \\			% needs transcriptome data
NucleotideContent	\quad Nucleotide Content \\							% needs checking for what it should do
reversecomplement	\quad Reverse Complement Anticodon \\
SetupRA	\quad Relative Synonymous Codon Usage \\

\subsubsection{Graphical functions}
plotOCU	\quad Plot of Optimality Score for Codons in a Single Protein \\

\subsubsection{3rd party functions}
movingAverage	\quad Calculating a Moving Average \\

\subsubsection{Obsolete functions}
countcodonfreq	\quad Count Codon Frequency \\

\subsubsection{Data Handling}
mylist	\quad My List of Nucleotide-Sequences and Aminoacid-Sequences \\
readfasta	\quad Processed tRNA Containing Fasta File Headers to Count Anticodons \\
readintronic	\quad Read tRNA Intronic Gene Count Statistics \\
readstats	\quad Read tRNA Gene Count Statistics \\
\hypertarget{function:rtRNAo}{readtRNAout}	\quad Read tRNA Gene Count Statistics \\
searchfafile	\quad Search and Return a Fasta File Name from Partial Name \\

\section{Sampledata}

\subsection{External Data}
Etef.sample.protein.fasta \quad Selectped Teff Protein sequences, matching selected CDS \\
Etef.sample.transcript.fasta \quad  Selected Teff CDS, closely related to rice sequences \\
\hypertarget{data:tRNAlist}{seq30469.out} \quad tRNAscan-SE 2.0 online output file for teff tRNA sequences is used as example imput for \hyperlink{function:rtRNAo}{readtRNAout} \\

\subsection{Datasets}
aa\_ac	\quad Aminoacids and Anticodons \\
GtRNAdb2species	\quad Viruses, Eukaryota, Archaea, Bacteria Species Names on GtRNAdb2 \\
Tef	\quad Optimal Codon Table for Eragostis tef Decided on tRNAscan-SE count of tRNA genes in sequenced genome \\
\hypertarget{data:veab}{veab}		\quad Viruses, Eukaryota, Archaea, Bacteria Species File Names on GtRNAdb2 \\
veabfa	\quad Viruses, Eukaryota, Archaea, Bacteria Species Fasta File Names on GtRNAdb2 \\




