\chapter{Outlook}
In this chapter we will give a brief overview on what Indices are meant to be implemented in the future, what results could be generated by applying them and what the future for this package looks like. 


\section{Codon Bias Indices}
There are dozens of codon bias indices that yet have to be implemented in the package. The original Darwin code has alredy defined some of the most wanted indices yet to be implemented. 
Here is an overview on the implemented, the ones which have a void function backbone and the non-implemented functions:

\begin{table}[tb]
\caption[Codon Bias Indices]{Codon Bias Indices}
\label{tab:CodonBiasIndices}
\centering
\begin{tabular}{cccc}
\toprule
	name 	& implemented & backbone & missing	  \\ 
\midrule
	Fop	& x	 & & \\
	CBI		& & x &	  \\
	B		& & x &   \\
	E		& & x &	  \\
\bottomrule 
\end{tabular}
\end{table}



The table~\ref{tab:CodonBiasIndices} is a floating table showing which indices have already been implemented in statanacoseq.

\section{Biological meaning and statistics}
The basic idea of a statistical analysis package for codon analysis is to combine the most frequently used codon indices in one software to evaluate which of them is best performing regarding to the availability (quality) and the biological properties for the given species. By putting it into the environement of the statistical open source program R following analysis steps as correlation to Gene Ontology (GO) tags can be done without having to deal with data handling to other software. The open structure of R codes also allows to use other implementation of algorithms already developed for R as we here used the functions RSCU from seqinr package for example. That allows other scientist to work with there own interpretation of new indices or better implementation of algorithms to calculate them. Furthermore, it is designed to adapt collaborators or third party code in the package, because every function has its own file and comes with the description of the function.
The basic biological problem to solve, as soon as all the major indices are implemented, is to find correlations of indices or clusters with other biological information such as gene expression (that is already part of the calculation of some indices), intracellular location, stress responses, enzymatic families, pathways, development stages and many more. One possible approach is to correlate the index properties to GO attributes and to find out wether there are indices that are significantly representing some of them, in order that a prediciton of the function or other properties of unknown proteins can be predicted.

\section{Package development}