\chapter{Outlook}
In this chapter we will give a brief overview on what Indices are meant to be implemented in the future, what results could be generated by applying them and what the future for this package looks like. 


\section{Codon Bias Indices}
There are dozens of codon bias indices that yet have to be implemented in the package. The original Darwin code has alredy defined some of the most wanted indices yet to be implemented. Among them are as example normalized translational efficiency (nTE) index or the tRNA adaptation index (tAI) \cite{Goodman2013}. \\
Here is an overview on the implemented, the ones which have a void function backbone and the non-implemented functions {tab:CodonBiasIndices}. The indices are taken from a review \cite{BinderCh13} and can be found in other reviews of codon indices \cite{Jansen2003}.

\begin{table}[tb]
\begin{footnotesize}
\caption[Codon Bias Indices]{Measures of codon bias}
\label{tab:CodonBiasIndices}
\centering
\begin{tabular}{cccc}
\toprule
	name 	& implemented & backbone & missing	  \\ 
\midrule
\textbf{Relative codon frequencies} \\
	$g_c$ / $g_{ac}$ global codon frequency	& x* & & 	 \\
	$f_{ac}$	 frequency of codon c encoding amino acid a & x* & &	 \\
	$r_{ac}$ relative synonymous codon usage (RSCU) for codon c and amino acid a	& x* & &	 \\
	$w_{ac}$	 relative adaptiveness & x & & 	 \\
\midrule
\textbf{Measures based on reference} \\
	Fop	Frequency of optimal codons & x	 & & \\
	\textbf{CBI}	Codon Bias Index	& & x &	 \\
	\textbf{B} / E	Codon Usage Bias (expression) 	& & & x  \\
	\textbf{CEC}	Codon-Enrichment Correlation 	& & & x	 \\
\midrule	
\textbf{Measures based on the geometric mean} \\
	\textbf{P}	Codon Preference 	& & & x	 \\
	\textbf{CAI}	Codon Adaptation Index	& x & &	 \\ % http://www.ncbi.nlm.nih.gov/pmc/articles/PMC340524/
	\textbf{rCAI}	 Relative Codon Adaptation Index	& & & x	 \\ % http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2880845/
	\textbf{RCB}	Relative Codon Usage Bias	& & & x	 \\
	RCA Relative codon adaptation	& & & x	 \\
	\textbf{FFT} Autocorrelation	& & & x	 \\
	\textbf{GCB}	Iterative approach to determining codon bias	& & & x	 \\
	\textbf{tAI}	tRNA adaptation index	& & & x	 \\
	nTE	Normalized translational efficiency	& & & x	 \\
\midrule	
\textbf{Measures based on deviation from an expected distribution} \\
	\textbf{CPB}	Codon-Preference Bias Measure	& & & x	 \\
	\textbf{MCB}	Maximum-likelihood Codon Bias	& & & x	 \\
	$B_a$ Bias of an individual amino acid	& & & x	 \\
	$\chi^2$ Scaled \textbf{chi squared} statistic		& & x &	 \\
	\textbf{Ew}	Weighted sum of relative entropy	& & & x	 \\
	\textbf{SCUO}	 Synonymous codon usage order	& & & x	 \\
\midrule	
\textbf{Measures focusing on tRNA interaction} \\
	\textbf{P1}	Influence of tRNA availability	& & & x	 \\
	P2	Bias for anticodonâ€“codon interactions of intermediary strength	& & & x	 \\
	\textbf{TPI}	tRNA-pairing index	& & & x	 \\
\midrule	
\textbf{Measures based on intrinsic properties of codon usage} \\
	\textbf{GC3s}	 Base composition at silent sites	& x* & &	 \\
	\textbf{Nc}	Effective number of codons	& x & &	 \\
	\textbf{MILC}	 Measure independent of length and composition	& & & x	 \\
	\textbf{MELP}	 MILC Expression Level Prediction	& & & x	 \\
	\textbf{ICDI}	 Intrinsic codon bias index	& & & x	 \\
	\textbf{HK}	Multivariate statistical method by Hey and Kliman	& & & x	 \\
	\textbf{S2str} Strength of mRNA secondary structure	& & & x	 \\
	\textbf{ER}	Evolutionary rate	& & & x	 \\
	\textbf{v(c)}	 Codon volatility	& & & x	 \\
	PLS	Partial least squares regression	& & & x	 \\
	\textbf{SUMBLE}	Synonymous codon usage bias maximum-likelihood estimation	& & & x	 \\
	\textbf{SEMPPR}	Stochastic evolutionary model of protein production rate	& & & x	 \\
	\midrule	
\textbf{Measures for total codon usage - genome wide} \\
	tot. Codon using		& & & x	 \\
	$D_{mean}$	\textbf{Mean} dissimilarity index 	& & & x	 \\
	\textbf{BC6}  base composition of 6mers \textit{(new)} 	& & & x	 \\
		\midrule	
\textbf{Indices measuring amino acid usage} \\
	GRAVY	Grand averages of hydropathy	& & & x	 \\
	AROMA	Aromaticity score	& & & x	 \\
	$k_s$	Rate of protein decay	& & & x	 \\
\bottomrule 

* needs proper implementation
\end{tabular}
\end{footnotesize}
\end{table}

The table~\ref{tab:CodonBiasIndices} shows which indices have already been implemented in statanacoseq.

To give those indices more power, attributes should be set to choose from different version of those indices that have been improoved over time. Test should be done with the results from the original papers to confirm the down-gradability of the index functions.


\section{Biological meaning and statistics}
The basic idea of a statistical analysis package for codon analysis is to combine the most frequently used codon indices in one software to evaluate which of them is best performing regarding to the availability (quality) and the biological properties for the given species. By putting it into the environement of the statistical open source program R following analysis steps as correlation to Gene Ontology (GO) tags can be done without having to deal with data handling to other software. The open structure of R codes also allows to use other implementation of algorithms already developed for R as we here used the functions RSCU from seqinr package for example. That allows other scientist to work with there own interpretation of new indices or better implementation of algorithms to calculate them. Furthermore, it is designed to adapt collaborators or third party code in the package, because every function has its own file and comes with the description of the function.\\
The basic biological problem to solve, as soon as all the major indices are implemented, is to find correlations of indices or clusters with other biological information such as gene expression (that is already part of the calculation of some indices), intracellular location, stress responses, enzymatic families, pathways, development stages and many more. One possible approach is to correlate the index properties to GO attributes and to find out wether there are indices that are significantly representing some of them, in order that a prediciton of the function or other properties of unknown proteins can be predicted.\\

\dots such GO enrichement analyses then may be directly used \dots \\

\section{Package development}
The future development of the package is heavily dependent of the scientific interest in a package that computes all the codon bias indices and to find the best correlation to the phenotypic attribute to be investigated on. 
Once a working packages is created it may be submitted to an official R package depository at \href{https://cran.r-project.org/doc/manuals/r-release/R-exts.html}{CRAN} or \href{https://www.bioconductor.org/developers/how-to/buildingPackagesForBioc/}{Bioconductor}.

